File: C++/1061-dating.cc
#include <iostream>
#include <string>
#include <array>
#include <cctype>
#include <iomanip>


const std::array<std::string, 7> kWeekName = {
    "MON", "TUE", "WED", "THU", "FRI", "SAT", "SUN"
};


int main() {

    std::string day;
    int hour, minute;

    std::string str1, str2;
    std::getline(std::cin, str1);
    std::getline(std::cin, str2);
    bool day_is_found = false;
    for (auto iter1 = str1.begin(), iter2 = str2.begin();
         iter1 != str1.end() && iter2 != str2.end();
         ++iter1, ++iter2)
    {
        if (*iter1 == *iter2) {
            if (!day_is_found) {
                if (*iter1 <= 'G' && *iter1 >= 'A') {
                    day = kWeekName[*iter1 - 'A'];
                    day_is_found = true;
                }
            } else {
                if (*iter1 <= 'N' && *iter1 >= 'A') {
                    hour = *iter1 - 'A' + 10;
                    break;
                } else if (std::isdigit(*iter1)) {
                    hour = *iter1 - '0';
                    break;
                }
            }
        }
    }
    std::string str3, str4;
    std::getline(std::cin, str3);
    std::getline(std::cin, str4);
    for (auto iter1 = str3.begin(), iter2 = str4.begin();
         iter1 != str3.end() && iter2 != str4.end();
         ++iter1, ++iter2)
    {
        if (*iter1 == *iter2 && std::isalpha(*iter1)) {
            minute = std::distance(str3.begin(), iter1);
        }
    }

    std::cout << day << ' ';
    std::cout << std::setfill('0');
    std::cout << std::setw(2) << hour << ':';
    std::cout << std::setw(2) << minute << '\n';
    std::cout << std::setfill(' ');

    return 0;
}

File: Python/1061-dating.py
import sys


letter_to_day = {
    'A': 'MON',
    'B': 'TUE',
    'C': 'WED',
    'D': 'THU',
    'E': 'FRI',
    'F': 'SAT',
    'G': 'SUN'
}


def decode(strings: str) -> str:
    day_is_found = False
    for first, second in zip(strings[0], strings[1]):
        if first == second:
            if day_is_found is False:
                if 'A' <= first <= 'G':
                    day = letter_to_day[first]
                    day_is_found = True
            else:
                if 'A' <= first <= 'N':
                    hour = ord(first) - ord('A') + 10
                    break
                elif first.isdigit():
                    hour = int(first)
                    break
    for i, (first, second) in enumerate(zip(strings[2], strings[3])):
        if first.isalpha() and second.isalpha() and first == second:
            minute = i
            break
    return f'{day} {hour:02d}:{minute:02d}'


if __name__ == '__main__':
    strings = sys.stdin.read().split()
    print(decode(strings))
